File: prototype/php/src/app/Domain/Escrow/PlatformMoney.php
<?php

declare(strict_types=1);

namespace App\Domain\Escrow;

use App\Domain\Money\Money;

/**
 * The whole platform's money in one place: escrow across every seller
 * ({@see LedgerBalances::total()}) beside what the platform kept and gave
 * back ({@see PlatformFees}). `/admin` and `/admin/accounting` show the same
 * six figures.
 */
final readonly class PlatformMoney
{
    private function __construct(
        public Money $held,
        public Money $available,
        public Money $paidOut,
        public Money $refunded,
        public Money $feesEarned,
        public Money $feesRefunded,
    ) {}

    public static function of(LedgerBalance $balance, PlatformFees $fees): self
    {
        return new self(
            $balance->held,
            $balance->available,
            $balance->paidOut,
            $balance->refunded,
            $fees->earned,
            $fees->refunded,
        );
    }
}

File: prototype/rails/src/app/models/platform_money.rb
# The platform's whole ledger balance beside what it earned and forgave in
# fees, and what went back to customers — the numbers /admin and
# /admin/accounting both show. Every figure is folded once, in a query
# whose cost does not grow with the number of sellers.
PlatformMoney = Data.define(:held, :available, :paid_out, :fees_earned, :fees_refunded, :refunded) do
  def self.fold
    balance = LedgerEntry.balance

    new(
      held: balance.held,
      available: balance.available,
      paid_out: balance.paid_out,
      fees_earned: Money.from_cents(Fulfillment.fees_earned_cents),
      fees_refunded: Money.from_cents(Fulfillment.fees_refunded_cents),
      refunded: Money.from_cents(Refund.sum(:amount_cents))
    )
  end
end
